In [4]:
import pandas as pd
import numpy as np
import keras
from keras.models import Sequential
from keras.layers import Dense
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [5]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [16]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

predictors_norm = (predictors - predictors.mean()) / predictors.std()
target_norm = (target - target.mean()) / target.std()

X=predictors_norm
y=target_norm

n_cols = predictors_norm.shape[1] # number of predictors

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)

## Build the Neural Network

In [17]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# build the model
model = regression_model()

## Fit the Model, Find the predictions, and Calculate the MSE

In [18]:
# fit the model
model.fit(X_train, y_train, epochs=50, verbose=0)


y_pred=model.predict(X_test)
MSE=mean_squared_error(y_test, y_pred)

print(MSE)

0.22830529449451584


## Repeat the process for 50 times

In [19]:
MSE = 0
MSEs = []

for i in range(0, 50):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3)
    model.fit(X_train, y_train, epochs=50, verbose=0)
    y_pred = model.predict(X_test)
    MSE = mean_squared_error(y_test, y_pred)
    print(MSE)
    MSEs.append(MSE)

107.47230887302803
82.96317361379593
79.57511719114086
69.57640196592145
69.0065015617111
54.18878939621051
55.08592346121889
55.49624452707913
55.10180896791697
45.501584044722215
52.89919668838656
44.662013507518886
54.58436323388208
50.03158239758512
47.74283874166382
44.53566733777086
41.153339078327065
42.83105479385983
43.00542002123305
47.249185209128505
49.97366129168863
46.08559727884218
56.94323761567017
43.13250002768376
38.7112347376352
41.47435532250548
36.61405619804112
46.1001318537691
46.76740945454481
55.22067971403817
48.03054367172385
39.29065228883423
46.87376895963467
42.81853326158861
44.78173658321323
45.62398195050693
38.954032704206064
47.717186415047294
45.92160338894054
34.936942195243226
37.41988792380715
41.10491548152954
47.21641533109039
49.190015089248824
41.37918784883766
52.64230453801666
39.80639970870839
46.8865432780901
33.63293464477477
39.22724968200919


In [20]:
MSEs = np.array(MSEs)
mean = np.mean(MSEs)
standard_deviation = np.std(MSEs)

print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

Mean: 49.342804261031425
Standard Deviation: 12.923391688540747


### While the mean and standard deviation of the non-normalized dataset are 66.59 and 25.92, the said parameters are respectively 49.34 and 12.92 for the normalized data. 

### Hence, both of the parameters decrease after the normalization.